### PR TITLE
fix(nuxt): provide node16-style type exports

### PR DESCRIPTION
**Pull request by danielroe**
### 🔗 Linked issue

https://github.com/nuxt/nuxt/issues/19606

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation or readme)
- [x] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [ ] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR makes some updates to Nuxt internals to improve support for Node16-style module resolution.

### 👉 Current status

<details>
<summary><strong>nuxt v3.2.3</strong></summary>

<table id="resolutions" class="">
<thead>
<tr>
<th></th>
<th><code>"nuxt"</code></th><th><code>"nuxt/config"</code></th><th><code>"nuxt/schema"</code></th><th><code>"nuxt/app"</code></th><th><code>"nuxt/package.json"</code></th>
</tr>
</thead>
<tbody>

<tr>
<td><code>node10</code></td>
<td>✅ </td><td>✅ </td><td>✅ </td><td>✅ </td><td>✅ (JSON)</td>
</tr>
<tr>
<td><code>node16</code> (from CJS)</td>
<td>❌ No types<br>⚠️ ESM (dynamic import only)</td><td>👺 Masquerading as ESM<br>⚠️ ESM (dynamic import only)</td><td>⚠️ ESM (dynamic import only)</td><td>❌ No types<br>⚠️ ESM (dynamic import only)</td><td>✅ (JSON)</td>
</tr>
<tr>
<td><code>node16</code> (from ESM)</td>
<td>❌ No types</td><td>✅ (ESM)</td><td>✅ (ESM)</td><td>❌ No types</td><td>✅ (JSON)</td>
</tr>
<tr>
<td><code>bundler</code></td>
<td>❌ No types</td><td>✅ </td><td>✅ </td><td>❌ No types</td><td>✅ (JSON)</td>
</tr>
</tbody></table>

</details><details>
<summary><strong>nuxt (edge)</strong></summary>

<table id="resolutions" class="">
<thead>
<tr>
<th></th>
<th><code>"nuxt3"</code></th><th><code>"nuxt3/config"</code></th><th><code>"nuxt3/schema"</code></th><th><code>"nuxt3/kit"</code></th><th><code>"nuxt3/app"</code></th><th><code>"nuxt3/package.json"</code></th>
</tr>
</thead>
<tbody>

<tr>
<td><code>node10</code></td>
<td>✅ </td><td>✅ </td><td>✅ </td><td>✅ </td><td>✅ </td><td>✅ (JSON)</td>
</tr>
<tr>
<td><code>node16</code> (from CJS)</td>
<td>❌ No types<br>⚠️ ESM (dynamic import only)</td><td>👺 Masquerading as ESM<br>⚠️ ESM (dynamic import only)</td><td>⚠️ ESM (dynamic import only)</td><td>⚠️ ESM (dynamic import only)</td><td>❌ No types<br>⚠️ ESM (dynamic import only)</td><td>✅ (JSON)</td>
</tr>
<tr>
<td><code>node16</code> (from ESM)</td>
<td>❌ No types</td><td>✅ (ESM)</td><td>✅ (ESM)</td><td>✅ (ESM)</td><td>❌ No types</td><td>✅ (JSON)</td>
</tr>
<tr>
<td><code>bundler</code></td>
<td>❌ No types</td><td>✅ </td><td>✅ </td><td>✅ </td><td>❌ No types</td><td>✅ (JSON)</td>
</tr>
</tbody></table>

</details><details>
<summary><strong>after this pr</strong></summary>

<table id="resolutions" class="">
<thead>
<tr>
<th></th>
<th><code>"nuxt"</code></th><th><code>"nuxt/config"</code></th><th><code>"nuxt/schema"</code></th><th><code>"nuxt/kit"</code></th><th><code>"nuxt/app"</code></th><th><code>"nuxt/package.json"</code></th>
</tr>
</thead>
<tbody>

<tr>
<td><code>node10</code></td>
<td>✅ </td><td>✅ </td><td>✅ </td><td>✅ </td><td>✅ </td><td>✅ (JSON)</td>
</tr>
<tr>
<td><code>node16</code> (from CJS)</td>
<td>⚠️ ESM (dynamic import only)</td><td>👺 Masquerading as ESM<br>⚠️ ESM (dynamic import only)</td><td>⚠️ ESM (dynamic import only)</td><td>⚠️ ESM (dynamic import only)</td><td>⚠️ ESM (dynamic import only)</td><td>✅ (JSON)</td>
</tr>
<tr>
<td><code>node16</code> (from ESM)</td>
<td>✅ (ESM)</td><td>✅ (ESM)</td><td>✅ (ESM)</td><td>✅ (ESM)</td><td>✅ (ESM)</td><td>✅ (JSON)</td>
</tr>
<tr>
<td><code>bundler</code></td>
<td>✅ </td><td>✅ </td><td>✅ </td><td>✅ </td><td>✅ </td><td>✅ (JSON)</td>
</tr>
</tbody></table>
</details>

See https://arethetypeswrong.github.io.

the remaining blockers for full support for Node16 format can be tracked in https://github.com/nuxt/nuxt/issues/19606.

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by dargmuesli** [comment]
minor notice: the issue mentioned is not linked properly because of a typo in "resolves" 🤫

File: packages/kit/package.json [hunk 1]
```diff
@@ -6,6 +6,14 @@
   "type": "module",
   "main": "./dist/index.mjs",
   "types": "./dist/index.d.ts",
+  "exports": {
+    ".": {
+      "types": "./dist/index.d.ts",
+      "import": "./dist/index.mjs",
+      "default": "./dist/index.mjs"
```
Comment (pi0) [inline packages/kit/package.json:13]: We might avoid default because this causes `require` to resolve into entry with wrong format. (kit is esm-only)

**Review by pi0 — COMMENTED** [review]
Nice work on this 🔥

File: packages/kit/package.json (hunk 1 above)
Comment (danielroe) [inline packages/kit/package.json:13]:
```suggestion
      "import": "./dist/index.mjs"
```

Comment (pi0) [reply packages/kit/package.json]: Backward compatibility. I think we can drop it for kit at least 👍🏼  (since it strictly requires ESM support)

**Comment by danielroe** [comment]
Okay, so _that's_ why I kept it in: https://github.com/nuxt/nuxt/actions/runs/4363710497/jobs/7630133013

**Comment by pi0** [comment]
Strange... It is vite or jiti? (jiti fallsback to `import` condition and we have enabled this behavior)

**Review by antfu — APPROVED** [review]

**Pull request merged**